### ISSUE TITLE
connect_socket has nonsensical behavior

### ISSUE THREAD

**Issue by tweenk** [civetweb/civetweb#756]
Based on my reading of the code, the internal function `connect_socket` has the following behavior:

1. Query `getaddrinfo()` for `AF_INET` addresses. Store the last returned address.
2. If no addresses were returned in step 1, query `AF_INET6` addresses. Store the last returned address.
3. Try to connect to the address that was stored.

This is doesn't make a lot of sense to me. A better behavior would be:
1. Query `getaddrinfo()` for `AF_UNSPEC` addresses.
2. Try to connect to each one until one of them accepts a connection or all fail.

**Comment by bel2125** [comment]
Yes, it indeed does behave like you wrote.
The reason why IPv4 is preferred to IPv6 is experience with lots of badly administrated IPv6 environments (some where completely "un-administrated" ... just, nobody turned IPv6 off and the operating system did *something*). In the code, IPv6 is used only when there is no IPv4.

Although, I have to admit this experience with IPv6 is already some years ago.
So, maybe by now there is no longer need to avoid IPv6 this way.

**Comment by tweenk** [comment]
Right, that is somewhat reasonable, although I think it should try both IPv4 and IPv6 networks. I know that internal networks at large cloud companies do not have IPv4 at all, because the network has too many hosts to fit in the IPv4 local network address space, and in such cases the library may be unable to resolve `localhost`.

The bigger problems are that the code will only attempt to connect to the last returned address for each IP protocol version, and that it overwrites the port number in the returned address instead of passing it to `getaddrinfo()`.

**Comment by bel2125** [comment]
Really, the 16 million IPs of 10.x.x.x are not enough? Or you mean they all need external IPs in the "internal" network? 

Anyway, in a big company there is probably more than one network admin knowing what IPv6 is and how to use it ... I had the issues more from home users (maybe small companies outside the IT sector) that copied the standalone (portable) executable to their windows PCs (with the operating system configuration right from the consumer electronics store) ... and nothing worked if IPv6 was enabled.

Do you have a suggestion (code snippet) to test?
